### DISCUSSION TITLE
How to run an exported protocol

### DISCUSSION THREAD

**Discussion by DanRothwell** [ukrobotics/DispenseLibDotNet#19] [Q&A]
I have a use case where for each plate, I need to fetch the protocol from a third-party system which dynamically constructs it for the plate. I need to be able to run a protocol from an exported file, either by saving it and running it via its protocol ID, or by running the exported file directly. Is this possible?

I see a command CompileDispense but that doesn't give a protocol ID.

**Comment by mcounsell** [comment]
You don't have to run via a protocol ID. You can write your code to create your own ProtocolData instance from scratch and pipe that down to the D2.  See the code (all open source) for how this is done. 

The protocol ID has no meaning to the D2 firmware, it is only used to support your driver when you need to support a user creating a protocol via our user interface and then allowing this protocol to be run from your driver. Think of the ID as a filepath, an uri or a google doc share link.... it is just a readonly reference to allow you to say "D2 - run this ".  

If you have some LIMS or other source of your protocol data then you can ignore the D2's user interface and protocols and simply create your own instance of ProtocolData. See the code , this is really just a ProtocolWell instances where each well has ValveCommand instances (for valve 1 and 2 and the volume to dispense ).  So if you want to roll your own just-in-time protocols start by looking at the method RunDispense(string protocolId, string plateTypeGuid) in class D2Controller... use this as a reference, copy these few lines of code into your driver.... also see

        public List<string> CompileDispense(
            ActiveCalibrationData activeCalibrationData, 
            ProtocolData protocolData, 
            PlateTypeData plateType)

Note this takes an instance of ProtocolData... you can create this your self in your code... it does not need to come from a protocol ID or or cloud DB.

Please reply with anything that is unclear and I will try to clarify as this is a common request and could help others...


TLDR:-   Super easy to do what you request, just create your own instance of ProtocolData and have the code pipe this down to D2... ignore the ID when you roll your own protocol as it has no purpose in this use case.

**Reply by DanRothwell** [reply]
Thanks for the reply. What I was hoping for was access to the methods that are used to import/export protocols, plate types and calibrations. That way we don't have to duplicate the file parsing logic or update it in the event of future changes to the file formats.

Protocols are the main target but the reason I mention plate types and calibrations too is that we often isolate our systems from the internet for security reasons. Having the ability to run the above method from local data would be necessary in that case, and doing so from local exports of the online data would be convenient for users.
